File: solutions/0052-n-queens-ii/n-queens-ii.py
# The n-queens puzzle is the problem of placing n queens on an n×n chessboard such that no two queens attack each other.
#
#
#
# Given an integer n, return the number of distinct solutions to the n-queens puzzle.
#
# Example:
#
#
# Input: 4
# Output: 2
# Explanation: There are two distinct solutions to the 4-queens puzzle as shown below.
# [
#  [".Q..",  // Solution 1
#   "...Q",
#   "Q...",
#   "..Q."],
#
#  ["..Q.",  // Solution 2
#   "Q...",
#   "...Q",
#   ".Q.."]
# ]
#
#


class Solution:
    
    def totalNQueens(self, n: int) -> int:
        self.res = 0
        board = [['.' for _ in range(n)] for _ in range(n)]
        self.solve(board, 0)
        return self.res
    
    def solve(self, board, cur_col):
        n = len(board)
        if cur_col == n:
            self.res += 1
            return True
        for row in range(n):
            if self.isvalid(board, row, cur_col):
                board[row][cur_col] = 'Q'
                self.solve(board, cur_col + 1)
                board[row][cur_col] = '.'
        return False

    def isvalid(self, board, r, c):
        # we have already ensured only one Queen in each column
        # only need to check rows
        n = len(board)
        for j in range(n):
            if board[r][j] == 'Q':
                return False
        i, j = r, c
        while i >= 0 and j >= 0:
            if board[i][j] == 'Q':
                return False
            i -= 1
            j -= 1
            
        i, j = r, c
        while i < n and j >= 0:
            if board[i][j] == 'Q':
                return False
            i += 1
            j -= 1
        return True

File: solutions/n-queens-ii.cpp
// The n-queens puzzle is the problem of placing n queens on an n x n chessboard such that no two queens attack each other.
//
// Given an integer n, return the number of distinct solutions to the n-queens puzzle.
//
//  
// Example 1:
//
//
// Input: n = 4
// Output: 2
// Explanation: There are two distinct solutions to the 4-queens puzzle as shown.
//
//
// Example 2:
//
//
// Input: n = 1
// Output: 1
//
//
//  
// Constraints:
//
//
// 	1 <= n <= 9
//
//


class Solution {
public:
    int n, ans;
    int totalNQueens(int n_) {
        n = n_;
        vector<string> s;
        for (int i = 0; i < n; i++){
            string tmp = "";
            for (int j = 0; j < n; j++){
                tmp += ".";
            }
            s.push_back(tmp);
        }
        solve(0, s);
        return ans;
    }
    bool check(int r, int c, vector<string>& s){
        for (int j = 0; j < n; j++){
            if (s[r][j] == 'Q') return false;
        }
        for (int i = 0; i < n; i++){
            if (s[i][c] == 'Q') return false;
        }
        for (int i = 0; i < n; i++){
            for (int j = 0; j < n; j++){
                if ((i + j == r + c) || (i - j == r - c)){
                    if (s[i][j] == 'Q')
                        return false;
                }
            }
        }
        return true;
    }
    void solve(int r, vector<string>& s){
        if (r >= n) {
            ans += 1;
            return ;
        }
        // int r = row(i), c = col(i);
        for (int c = 0; c < n; c++){
            if (check(r, c, s)){
                s[r][c] = 'Q';
                solve(r + 1, s);
                s[r][c] = '.';
            }
        }
        return ;
    }
};
